StorageOS configurable device directory and mount options

croomes opened the pull request:
**What this PR does / why we need it**:
This allows StorageOS volumes to be mounted when the kubelet is running in a container and we are unable to use the default device location (/var/lib/storageos/volumes).  With this PR, the node's device location is requested via the StorageOS api, falling back to the current behaviour if not configured.  The node's device location can be supplied as an environment variable (DEVICE_DIR) to the StorageOS container.  This is backwards-compatible and no changes are needed to existing deployments.

The PR also allows Mount options to be set for StorageOS volumes in the same way they're enabled for other volume plugins.

The StorageOS API dependency was updated to the latest version, but no functionality changes besides adding the DeviceDir property to the Controller object.

There is also a small refactor of the loopback device handling code in storageos_utils.go to capture stderr output.

**Release note**:
```release-note
StorageOS volume plugin updated to support mount options and environments where the kubelet runs in a container and the device location should be specified.
```

Not sure why godep changed the comments of unrelated packages in Godeps.json...

/sig storage

jsafrane commented on pkg/volume/storageos/storageos_util.go:
```diff
@@ -323,27 +346,30 @@ func getLoopDevice(path string, exec mount.Exec) (string, error) {
 
 	args := []string{"-j", path}
-	out, err := exec.Run(losetupPath, args...)
+	cmd := exec.Command(losetupPath, args...)
```
Please add a comment that you intentionally run where kubelet is (e.g. in kubelet's container).

croomes replied:
@jsafrane thanks for the review, but quick question as I noticed that I reverted the VolumeHost.GetExec() change you made without fully understanding it.

While we expect the mount utilities (incl `losetup`) to run where the kubelet runs, we don't require it.  We just need the device and mount to be available in the correct place and where the kubelet can see them.

Unless you can think of any gotchas or I'm misunderstanding, I'll revert this bit.

Also fyi: the Block Volume support loop device code seems to have used this as a base, and doesn't use VolumeHost.GetExec().  Not sure if this was intentional, and if it was do I need the same here? https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/util_linux.go.  Maybe @mtanino can advise?

jsafrane replied:
In 1.9 there is a possibility to run mount utilities (mount.nfs, iscsiadm, mkfs.btrfs, ..., losetup) in a dedicated container instead of the host (which can be a minimal one like CoreOS) or in the container where kubelet runs (which can have just kubelet). The setup is quite elaborate and it is dedicated only for testing in our e2e tests, which run quite minimal GCI image (but it has losetup).   `VolumeHost.GetExec()` is used to execute things in these containers with mount utilities instead of on the host. You probably should use this interface for losetup too, just in case someone has really minimal OS without it.
 
In addition, it would be great if you could prepare container(s) with the server parts of StorageOS and e2e tests that would deploy it and test the plugin against. This way, we could test our plugin and make sure there are no regressions as the plugin evolves.

croomes replied:
Thanks, I'll revert this bit.  We're working on the e2e tests separately.  I'm hoping we can get a PR  for them in the next few weeks.

jsafrane replied:
I still see you use `exec.Command` directly, did you plan to revert it back to `mount.Exec` interface?

jsafrane commented on pkg/volume/storageos/storageos_util.go:
```diff
@@ -323,27 +346,30 @@ func getLoopDevice(path string, exec mount.Exec) (string, error) {
-	out, err := exec.Run(losetupPath, args...)
+func makeLoopDevice(path string) (string, error) {
+	args := []string{"-f", "-P", "--show", path}
+	cmd := exec.Command(losetupPath, args...)
```
Please add a comment that you intentionally run where kubelet is (e.g. in kubelet's container).

jsafrane commented on pkg/volume/storageos/storageos_util.go:
```diff
@@ -323,27 +346,30 @@ func getLoopDevice(path string, exec mount.Exec) (string, error) {
+func removeLoopDevice(device string) error {
 	args := []string{"-d", device}
-	out, err := exec.Run(losetupPath, args...)
+	cmd := exec.Command(losetupPath, args...)
```
Please add a comment that you intentionally run where kubelet is (e.g. in kubelet's container).

jsafrane commented:
/ok-to-test

jsafrane commented:
/retest

jsafrane commented:
flake: #58975

cblecker commented:
/retest

croomes commented:
@jsafrane reverted the exec changes, but just noticed the Godeps conflict, will resolve tomorrow.

croomes committed:
Update storageos api dependency to 0.3.4

croomes committed:
StorageOS support for containerized kubelet and mount options

This change queries the StorageOS API to retrieve the volume
device dir when a volume is attached, rather than relying on
the default /var/lib/storageos/volumes directory.  This allows
all or some nodes to have kubelet running in a container with
the StorageOS volume path mounted in, which is required for
Azure ACS support.

Volume mount options are also supported.

croomes commented:
@jsafrane  I think this is good to go now, PTAL - thanks

jsafrane commented:
/lgtm

jsafrane commented:
/assign @thockin 
for vendor approval

jsafrane commented:
/approve no-issue
/retest

croomes commented:
/test pull-kubernetes-e2e-kubeadm-gce-canary

croomes commented:
/test pull-kubernetes-e2e-kubeadm-gce-canary

croomes commented:
/test pull-kubernetes-e2e-kubeadm-gce-canary

croomes commented:
/test pull-kubernetes-e2e-kubeadm-gce-canary

thockin commented:
In the future, please file bugs so we can xref them in PRs.

/lgtm
/approve no-issue

croomes commented:
/status approved-for-milestone

croomes commented:
@saad-ali are you able to help get this into 1.10 release?  Was hoping it would have gone in automatically...

jsafrane commented:
Don't worry, it's merged into master so it's going to be part of 1.10. It splits from master in upcoming week(s).

croomes commented:
Fantastic, thanks for letting me know!
